src/NuGet.Server.Core/Infrastructure/ServerPackageCache.cs

```diff
@@ -0,0 +1,261 @@
+                        }
+                        else
+                        {
+                            throw;
+                        }
+                    }
+                }
+            }
+            finally
+            {
+                _syncLock.ExitWriteLock();
+            }
+        }
+
+        public bool HasPackages()
+        {
+            _syncLock.EnterReadLock();
+            try
+            {
+                return _packages.Count > 0;
+            }
+            finally
+            {
+                _syncLock.ExitReadLock();
+            }
+        }
+
+        public IEnumerable<ServerPackage> GetAll()
+        {
+            _syncLock.EnterReadLock();
+            try
+            {
+                return _packages.ToList();
+            }
+            finally
+            {
+                _syncLock.ExitReadLock();
+            }
+        }
+
+        public void Remove(string id, SemanticVersion version, bool enableDelisting)
+        {
+            _syncLock.EnterWriteLock();
+            try
+            {
+                if (enableDelisting)
+                {
+                    foreach (var package in _packages.Where(p => IsMatch(p, id, version)))
+                    {
+                        package.Listed = false;
+                    }
+                }
+                else
+                {
+                    _packages.RemoveWhere(p => IsMatch(p, id, version));
+                }
+
+                UpdateLatestVersions(_packages.Where(p => IsMatch(p, id)));
+
+                _isDirty = true;
+            }
+            finally
+            {
+                _syncLock.ExitWriteLock();
+            }
+        }
+
+        public void Add(ServerPackage entity)
+        {
+            _syncLock.EnterWriteLock();
+            try
+            {
+                _packages.Remove(entity);
+                _packages.Add(entity);
+
+                UpdateLatestVersions(_packages.Where(p => IsMatch(p, entity.Id)));
+
+                _isDirty = true;
+            }
+            finally
+            {
+                _syncLock.ExitWriteLock();
+            }
+        }
+
+        public void StoreRange(IEnumerable<ServerPackage> entities)
+        {
+            _syncLock.EnterWriteLock();
+            try
+            {
+                foreach (var entity in entities)
+                {
+                    _packages.Remove(entity);
+                    _packages.Add(entity);
+                }
+
+                UpdateLatestVersions(_packages);
+
+                _isDirty = true;
+            }
+            finally
+            {
+                _syncLock.ExitWriteLock();
+            }
+        }
+
+        private static void UpdateLatestVersions(IEnumerable<ServerPackage> packages)
+        {
+            // TODO: this should respect the Listed property on the package.
+            // https://github.com/NuGet/NuGetGallery/issues/3434
```

Comment:
I think the benefit of having TODO is it allows easy searching for questionable areas in the code without dependency on bug tracker.